File: p095m/unique_bst2.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
const Tree = require('binary_tree');
var generateTrees = function (n) {
  /**
   * @param {number} s: start
   * @param {number} e: end
   * @return {TreeNode[]}
   */
  const recTrees = function (s, e) {
    if (s > e) return [null];
    let trees = [];
    for (let root = s; root <= e; ++root) {
      let ltrees = recTrees(s, root - 1);
      for (let left of ltrees) {
        let rtrees = recTrees(root + 1, e);
        for (let right of rtrees) {
          let node = new Tree.TreeNode(root);
          node.left = left;
          node.right = right;
          trees.push(node);
        }
      }
    }
    return trees;
  };
  return recTrees(1, n);
};

// TESTS
[1, 2, 3, 4].forEach((n) => {
  console.log(
    'Unique BST of n =',
    n,
    '->',
    generateTrees(n).map(Tree.serialize)
  );
});

File: p095m/unique_bst2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import List, Optional
from functools import lru_cache
from local_packages.binary_tree import TreeNode


class Solution:
    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        @lru_cache
        def trees(first: int, last: int) -> List[Optional[TreeNode]]:
            return [
                TreeNode(root, left, right)
                for root in range(first, last + 1)
                for left in trees(first, root - 1)
                for right in trees(root + 1, last)
            ] or [None]

        return trees(1, n)


# TESTS
for n in range(1, 5):
    sol = Solution()
    actual = sol.generateTrees(n)
    print("Unique BST of n =", n, "->", [TreeNode.serialize(t) for t in actual])
